packages/ref-core/src/cmip_ref_core/providers.py

```diff
@@ -86,3 +116,233 @@ def get(self, slug: str) -> Metric:
+            with conda_exe.open(mode="wb") as file:
+                file.write(response.content)
+            conda_exe.chmod(stat.S_IRWXU)
+            logger.info("Successfully installed conda.")
+        elif update:
+            logger.info("Updating conda")
+            subprocess.run([str(conda_exe), "self-update"], check=True)  # noqa: S603
+            logger.info("Successfully updated conda")
+
+        return conda_exe
+
+    def get_conda_exe(self, update: bool = False) -> Path:
+        """
+        Get the path to a conda executable.
+        """
+        if self._conda_exe is None:
+            self._conda_exe = self._install_conda(update)
+        return self._conda_exe
+
+    def get_environment_file(self) -> AbstractContextManager[Path]:
+        """
+        Return a context manager that provides the environment file as a Path.
+        """
+        # Because providers are instances, we have no way of retrieving the
+        # module in which they are created, so get the information from the
+        # first registered metric instead.
+        metrics = self.metrics()
+        if len(metrics) == 0:
+            msg = "Unable to determine the provider module, please register a metric first."
+            raise ValueError(msg)
+        module = metrics[0].__module__.split(".")[0]
+        lockfile = importlib.resources.files(module).joinpath("requirements").joinpath("conda-lock.yml")
+        return importlib.resources.as_file(lockfile)
+
+    @property
+    def env_path(self) -> Path:
+        """
+        A unique path for storing the conda environment.
+        """
+        with self.get_environment_file() as file:
+            suffix = hashlib.sha1(file.read_bytes(), usedforsecurity=False).hexdigest()
+        return self.prefix / f"{self.slug}-{self.version}-{suffix}"
+
+    def create_env(self) -> None:
+        """
+        Create a conda environment.
+        """
+        if self.env_path.exists():
+            logger.info(f"Environment at {self.env_path} already exists, skipping.")
+            return
+
+        with self.get_environment_file() as file:
+            cmd = [
+                f"{self.get_conda_exe(update=True)}",
+                "create",
+                "--yes",
+                "--file",
+                f"{file}",
+                "--prefix",
+                f"{self.env_path}",
+            ]
+            logger.debug(f"Running {' '.join(cmd)}")
+            subprocess.run(cmd, check=True)  # noqa: S603
+
+    def run(self, cmd: Iterable[str]) -> None:
+        """
+        Run a command.
+
+        Parameters
+        ----------
+        cmd :
+            The command to run.
+
+        """
+        self.create_env()
+
+        cmd = [
+            f"{self.get_conda_exe(update=False)}",
+            "run",
+            "--prefix",
+            f"{self.env_path}",
+            *cmd,
+        ]
+        logger.info(f"Running {' '.join(cmd)}")
+        subprocess.run(cmd, check=True)  # noqa: S603
```

Comment:
We might want to capture that subprocess output at some point. Although the executor might be a more consistent place to capture logs of the entire execution process

Reply:
For ESMValTool, the logging messages are saved in its output directory. Stdout/stderr is lost though.

Reply:
stdout/stderr comes through to the console when running an ESMValTool metric. We might want to suppress that unless we are at a debug log level as it is extremely noisy when running all providers

Everyone does it slightly different so we should try find common pattern that works for all. I see that you have two different log files. There is a field in the CMEC bundles for the log output, but we don't do anything with that field at the moment and logs aren't automatically copied to the "results" directory.